### PR TITLE
Support min allowed app version

### LINKED ISSUE #445: Add support for min allowed app version
- [x] Pull `min_allowed_app_version` server setting during sync
- [x] Allow sync of already collected data but log user out after sync if current app version < `min_allowed_app_version`
- [x] Set flag after sync to prevent login after the forced logout above

### PR DESCRIPTION
**Pull request by vincent-karuri**
Closes #445

### CODE REVIEW

**Comment by keymane** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- opensrp-app/src/main/java/org/smartregister/view/activity/LoginActivity.java  1
- opensrp-app/src/main/java/org/smartregister/login/presenter/BaseLoginPresenter.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-core/pullRequest?prid=5144603&bid=16898740)

**Comment by keymane** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- opensrp-app/src/main/java/org/smartregister/login/presenter/BaseLoginPresenter.java  1
- opensrp-app/src/test/java/org/smartregister/presenter/TestLoginPresenter.java  1
- opensrp-app/src/main/java/org/smartregister/view/activity/LoginActivity.java  1
- opensrp-app/src/test/java/org/smartregister/presenter/BaseLoginPresenterTest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-core/pullRequest?prid=5144603&bid=16898740)

**Comment by keymane** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- opensrp-app/src/main/java/org/smartregister/login/presenter/BaseLoginPresenter.java  1
- opensrp-app/src/test/java/org/smartregister/presenter/BaseLoginPresenterTest.java  1
- opensrp-app/src/main/java/org/smartregister/view/activity/LoginActivity.java  1
- opensrp-app/src/test/java/org/smartregister/presenter/TestLoginPresenter.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-core/pullRequest?prid=5144603&bid=16898740)

**Comment by keymane** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- opensrp-app/src/test/java/org/smartregister/presenter/BaseLoginPresenterTest.java  1
- opensrp-app/src/test/java/org/smartregister/presenter/TestLoginPresenter.java  1
- opensrp-app/src/main/java/org/smartregister/login/presenter/BaseLoginPresenter.java  1
- opensrp-app/src/main/java/org/smartregister/view/activity/LoginActivity.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-core/pullRequest?prid=5144603&bid=16898740)

**Comment by keymane** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- opensrp-app/src/test/java/org/smartregister/util/SyncUtilsTest.java  1
- opensrp-app/src/test/java/org/smartregister/presenter/BaseLoginPresenterTest.java  1
- opensrp-app/src/main/java/org/smartregister/view/activity/LoginActivity.java  1
- opensrp-app/src/test/java/org/smartregister/presenter/TestLoginPresenter.java  1
- opensrp-app/src/main/java/org/smartregister/login/presenter/BaseLoginPresenter.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-core/pullRequest?prid=5144603&bid=16898740)

**Comment by keymane** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- opensrp-app/src/test/java/org/smartregister/util/SyncUtilsTest.java  1
- opensrp-app/src/test/java/org/smartregister/presenter/TestLoginPresenter.java  1
- opensrp-app/src/test/java/org/smartregister/presenter/BaseLoginPresenterTest.java  1
- opensrp-app/src/main/java/org/smartregister/login/presenter/BaseLoginPresenter.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-core/pullRequest?prid=5144603&bid=16898740)

**Comment by keymane** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- opensrp-app/src/test/java/org/smartregister/presenter/TestLoginPresenter.java  1
- opensrp-app/src/main/java/org/smartregister/login/presenter/BaseLoginPresenter.java  1
- opensrp-app/src/test/java/org/smartregister/presenter/BaseLoginPresenterTest.java  1
- opensrp-app/src/test/java/org/smartregister/util/SyncUtilsTest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-core/pullRequest?prid=5144603&bid=16898740)

File: opensrp-app/src/main/java/org/smartregister/sync/intent/SyncIntentService.java [hunk 1]
```diff
@@ -79,12 +82,16 @@ private void doSync() {
             if (hasValidAuthorization || !CoreLibrary.getInstance().getSyncConfiguration().disableSyncToServerIfUserIsDisabled()) {
                 pushToServer();
             }
-            if (hasValidAuthorization) {
-                pullECFromServer();
-            } else {
+
+            if (!syncUtils.isAppVersionAllowed()) {
+                AllSettings settingsRepository = org.smartregister.Context.getInstance().allSettings();
+                settingsRepository.put(IS_APP_VERSION_ALLOWED, "false");
```
Comment (githengi) [inline opensrp-app/src/main/java/org/smartregister/sync/intent/SyncIntentService.java]: This should be after `if(!hasValidAuthorization)` as its more efficient as `hasValidAuthorization` is already known.

Not sure why  `IS_APP_VERSION_ALLOWED` is being re-saved when its used to determine `syncUtils.isAppVersionAllowed()`

File: opensrp-app/src/main/java/org/smartregister/util/SyncUtils.java [hunk 2]
```diff
@@ -84,4 +96,39 @@ public void logoutUser() {
...
+
+    private int extractMinAllowedAppVersion(String setting) {
+        int minAllowedAppVersion = 0;
+        try {
+            JSONArray settings = new JSONObject(setting).optJSONArray(SETTINGS);
+            for (int i = 0; i < settings.length(); i++) {
+                JSONObject currSettingObj = settings.optJSONObject(i);
+                String currKey = currSettingObj.optString(KEY);
+                if (MIN_ALLOWED_APP_VERSION.equals(currKey)) {
+                    minAllowedAppVersion = currSettingObj.optInt(VALUE, minAllowedAppVersion);
```
Comment (githengi) [inline opensrp-app/src/main/java/org/smartregister/util/SyncUtils.java]: maybe return from loop here

Comment (vincent-karuri) [reply opensrp-app/src/main/java/org/smartregister/util/SyncUtils.java]: Done.

File: opensrp-app/src/main/java/org/smartregister/sync/intent/SyncIntentService.java (hunk 1 above)
Comment (vincent-karuri) [reply opensrp-app/src/main/java/org/smartregister/sync/intent/SyncIntentService.java]: Fixed the first concern.

I'm not sure I understand the second concern. An upgrade can only be done by uninstalling the current app or by clearing data on the phone and upgrading with a new apk. So the `IS_APP_VERSION_ALLOWED` will definitely be recomputed.

Once a user is logged out, they cannot log back in until they upgrade the app by following one of the two options above.

**Comment by keymane** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- opensrp-app/src/test/java/org/smartregister/presenter/TestLoginPresenter.java  1
- opensrp-app/src/test/java/org/smartregister/presenter/BaseLoginPresenterTest.java  1
- opensrp-app/src/main/java/org/smartregister/view/activity/LoginActivity.java  1
- opensrp-app/src/main/java/org/smartregister/login/presenter/BaseLoginPresenter.java  1
- opensrp-app/src/test/java/org/smartregister/util/SyncUtilsTest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-core/pullRequest?prid=5144603&bid=16898740)

File: opensrp-app/src/main/java/org/smartregister/sync/intent/SyncIntentService.java (hunk 1 above)
Comment (githengi) [reply opensrp-app/src/main/java/org/smartregister/sync/intent/SyncIntentService.java]: Its possible to upgrade without uninstall of the current app and this is the upgrade recommended for the OpenSRP.

The concert is valid when the APK is upgraded

Comment (vincent-karuri) [reply opensrp-app/src/main/java/org/smartregister/sync/intent/SyncIntentService.java]: Fixed.

File: opensrp-app/src/main/java/org/smartregister/AllConstants.java [hunk 3]
```diff
@@ -450,4 +450,15 @@
 
     }
 
+    public interface FORCED_LOGOUT {
+        String MIN_ALLOWED_APP_VERSION_SETTING = "min_allowed_app_version_setting";
+        String MIN_ALLOWED_APP_VERSION = "min_allowed_app_version";
+        String IS_APP_VERSION_ALLOWED = "is_app_version_allowed";
```
Comment (githengi) [inline opensrp-app/src/main/java/org/smartregister/AllConstants.java]: This is no longer used

File: opensrp-app/src/main/java/org/smartregister/sync/intent/SyncIntentService.java [hunk 4]
```diff
@@ -31,6 +32,8 @@
 
 import timber.log.Timber;
 
+import static org.smartregister.AllConstants.FORCED_LOGOUT.IS_APP_VERSION_ALLOWED;
```
Comment (githengi) [inline opensrp-app/src/main/java/org/smartregister/sync/intent/SyncIntentService.java]: import not used

**Comment by keymane** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- opensrp-app/src/test/java/org/smartregister/presenter/BaseLoginPresenterTest.java  1
- opensrp-app/src/test/java/org/smartregister/presenter/TestLoginPresenter.java  1
- opensrp-app/src/main/java/org/smartregister/view/activity/LoginActivity.java  2
- opensrp-app/src/main/java/org/smartregister/login/presenter/BaseLoginPresenter.java  1
- opensrp-app/src/test/java/org/smartregister/TestUtils.java  1
- opensrp-app/src/test/java/org/smartregister/util/SyncUtilsTest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-core/pullRequest?prid=5144603&bid=16898740)

**Comment by keymane** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- opensrp-app/src/test/java/org/smartregister/TestUtils.java  1
- opensrp-app/src/main/java/org/smartregister/view/activity/LoginActivity.java  2
- opensrp-app/src/test/java/org/smartregister/presenter/BaseLoginPresenterTest.java  1
- opensrp-app/src/main/java/org/smartregister/login/presenter/BaseLoginPresenter.java  1
- opensrp-app/src/test/java/org/smartregister/presenter/TestLoginPresenter.java  1
- opensrp-app/src/test/java/org/smartregister/util/SyncUtilsTest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-core/pullRequest?prid=5144603&bid=16898740)

**Comment by vincent-karuri** [comment]
Tested by following these steps:

- Post settings with min allowed app version
- Update app with outdated version code - should logout 
- Update app version code to the min allowed - should allow login / sync
- Bump min allowed version to be > current by posting to server - should now logout app on sync
- Downgrade min allowed version to be < current app version code by posting to server - should allow login / sync
- Again, bump min allowed version to be > current app version code by posting to server  - should now logout app on sync
- Update app version code to be > min allowed - should allow login / sync

**Comment by keymane** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- opensrp-app/src/main/java/org/smartregister/view/activity/LoginActivity.java  2
- opensrp-app/src/test/java/org/smartregister/presenter/BaseLoginPresenterTest.java  1
- opensrp-app/src/main/java/org/smartregister/login/presenter/BaseLoginPresenter.java  1
- opensrp-app/src/test/java/org/smartregister/util/SyncUtilsTest.java  1
- opensrp-app/src/test/java/org/smartregister/TestUtils.java  1
- opensrp-app/src/test/java/org/smartregister/presenter/TestLoginPresenter.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-core/pullRequest?prid=5144603&bid=16898740)

**Comment by keymane** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- opensrp-app/src/test/java/org/smartregister/presenter/BaseLoginPresenterTest.java  1
- opensrp-app/src/test/java/org/smartregister/TestUtils.java  1
- opensrp-app/src/main/java/org/smartregister/login/presenter/BaseLoginPresenter.java  1
- opensrp-app/src/main/java/org/smartregister/view/activity/LoginActivity.java  2
- opensrp-app/src/test/java/org/smartregister/util/SyncUtilsTest.java  1
- opensrp-app/src/test/java/org/smartregister/presenter/TestLoginPresenter.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-core/pullRequest?prid=5144603&bid=16898740)

File: opensrp-app/src/main/java/org/smartregister/AllConstants.java (hunk 3 above)
Comment (vincent-karuri) [reply opensrp-app/src/main/java/org/smartregister/AllConstants.java]: Removed.

File: opensrp-app/src/main/java/org/smartregister/sync/intent/SyncIntentService.java (hunk 4 above)
Comment (vincent-karuri) [reply opensrp-app/src/main/java/org/smartregister/sync/intent/SyncIntentService.java]: Removed.

**Comment by keymane** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- opensrp-app/src/test/java/org/smartregister/util/SyncUtilsTest.java  1
- opensrp-app/src/main/java/org/smartregister/view/activity/LoginActivity.java  2
- opensrp-app/src/main/java/org/smartregister/login/presenter/BaseLoginPresenter.java  1
- opensrp-app/src/test/java/org/smartregister/TestUtils.java  1
- opensrp-app/src/test/java/org/smartregister/presenter/BaseLoginPresenterTest.java  1
- opensrp-app/src/test/java/org/smartregister/presenter/TestLoginPresenter.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-core/pullRequest?prid=5144603&bid=16898740)

**Comment by keymane** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- opensrp-app/src/test/java/org/smartregister/presenter/BaseLoginPresenterTest.java  1
- opensrp-app/src/main/java/org/smartregister/login/presenter/BaseLoginPresenter.java  1
- opensrp-app/src/main/java/org/smartregister/view/activity/LoginActivity.java  2
- opensrp-app/src/test/java/org/smartregister/presenter/TestLoginPresenter.java  1
- opensrp-app/src/test/java/org/smartregister/util/SyncUtilsTest.java  1
- opensrp-app/src/test/java/org/smartregister/TestUtils.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-core/pullRequest?prid=5144603&bid=16898740)

**Comment by keymane** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- opensrp-app/src/main/java/org/smartregister/view/activity/LoginActivity.java  2
- opensrp-app/src/test/java/org/smartregister/presenter/TestLoginPresenter.java  1
- opensrp-app/src/test/java/org/smartregister/util/SyncUtilsTest.java  1
- opensrp-app/src/test/java/org/smartregister/TestUtils.java  1
- opensrp-app/src/main/java/org/smartregister/login/presenter/BaseLoginPresenter.java  1
- opensrp-app/src/test/java/org/smartregister/presenter/BaseLoginPresenterTest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/OpenSRP/opensrp-client-core/pullRequest?prid=5144603&bid=16898740)

**Pull request merged**